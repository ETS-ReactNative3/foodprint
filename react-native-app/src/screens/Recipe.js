import { ActivityIndicator, Image, StyleSheet, Text, View, BackHandler } from 'react-native';
import React, { useState, useEffect, useCallback } from 'react';
import { widthPercentageToDP as percentageWidth, heightPercentageToDP as percentageHeight } from 'react-native-responsive-screen';
import { Button, Input } from 'react-native-elements';
import { gql } from 'apollo-boost';
import { useMutation } from '@apollo/react-hooks';
import Snackbar from 'react-native-snackbar';
import { useNetInfo } from '@react-native-community/netinfo';

const POST_RECIPE_MUTATION = gql`
mutation($input: String!) {
  postRecipe(name: $input) {
     name
     carbonFootprintPerKg
     imageUrl
     ingredients {
        ingredient
        amountKg
        carbonFootprintPerKg
     }
     sourceUrl
  }
}`;

const Recipe = ({ navigation, route }) => {
  const netInfo = useNetInfo();
  const [input, setInput] = useState('');
  const [postRecipe, { loading: recipeLoading, error: recipeError, data: recipeData }] = useMutation(POST_RECIPE_MUTATION);

  useEffect(() => {
    if (netInfo.details !== null && !netInfo.isConnected) {
      Snackbar.show({
        text: 'No internet connection, you can\'t do this! ðŸ¤£',
        duration: Snackbar.LENGTH_INDEFINITE,
      });
    } else if (netInfo.isConnected) {
      Snackbar.dismiss();
    }
  }, [netInfo]);

  BackHandler.addEventListener('hardwareBackPress', () => {
    Snackbar.dismiss();
  });

  const handleSubmit = async () => {
    try {
      await postRecipe({ variables: { input } });
    } catch (err) {
      console.warn({ err });
    }
  };

  useEffect(() => {
    if (route.params && route.params.recipeUrl) {
      setInput(route.params.recipeUrl);
    }
    // eslint-disable-next-line react-hooks/exhaustive-deps
  }, []);

  useEffect(() => {
    if (!recipeError) {
      return;
    }
    Snackbar.show({
      text: 'Oops, something went wrong :(',
      duration: Snackbar.LENGTH_SHORT,
    });
  }, [recipeError]);


  useEffect(() => {
    if (recipeData && recipeData.postRecipe.carbonFootprintPerKg === null) {
      Snackbar.show({
        text: "Oops, we couldn't get the footprint for this recipe",
        duration: Snackbar.LENGTH_SHORT,
      });
    }
  }, [recipeData]);


  useEffect(() => {
    if (recipeData && recipeData.postRecipe.carbonFootprintPerKg != null) {
      navigation.navigate('Feedback', {
        recipeMeal: {
          uri: recipeData.postRecipe.imageUrl,
          score: recipeData.postRecipe.carbonFootprintPerKg,
          description: recipeData.postRecipe.name,
        }, extraInfo: {
          recipeUrl: recipeData.postRecipe.sourceUrl,
          ingredients: recipeData.postRecipe.ingredients,
        }, loading: false,
      });
    }
  }
  );


  return (
    <View style={styles.container}>
      <Image
        source={require('../images/full-smiley.png')}
        style={styles.image}
        resizeMode="contain"
      />
      <Text style={styles.text}>
        Want to know the carbon footprint of a meal you want to cook? Write the name or paste the URL of the
        recipe in the following field!
       </Text>
      <Input
        value={input}
        containerStyle={styles.input}
        onChangeText={value => setInput(value)}
      />
      {(recipeLoading) ? (
        <ActivityIndicator style={styles.loading} />
      ) : (
          <Button
            title="GO"
            buttonStyle={styles.button}
            containerStyle={styles.buttonContainer}
            titleStyle={styles.buttonTitle}
            onPress={handleSubmit}
          />
        )}
    </View>
  );
};

const styles = StyleSheet.create({
  container: { flex: 1, justifyContent: 'center', alignItems: 'center', margin: percentageHeight('5%') },
  image: { height: percentageHeight('25%'), marginBottom: percentageHeight('5%') },
  text: { fontSize: percentageWidth('4%') },
  input: { marginVertical: percentageHeight('2%') },
  button: { backgroundColor: 'green', width: percentageWidth('30%'), height: 45 },
  buttonContainer: { marginVertical: percentageHeight('2%') },
  buttonTitle: { fontSize: percentageWidth('5%') },
  loading: { height: 45, marginVertical: percentageHeight('2%') },
});

export default Recipe;
