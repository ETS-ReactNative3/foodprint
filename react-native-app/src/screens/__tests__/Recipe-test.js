import React from 'react';
import renderer from 'react-test-renderer';
import Recipe, { POST_RECIPE_MUTATION } from '../Recipe';
import { MockedProvider } from '@apollo/react-testing';

const mockedNavigation = {};

const mockedRoute = {};

test('matches snapshot when carbon footprint success', async () => {
  const mockedResponses = [
    {
      request: {
        query: POST_RECIPE_MUTATION,
      },
      result: () => {
        return {
          data: {
            postRecipe: {
              carbonFootprintPerKg: 0.5,
            },
          },
        };
      },
    },
  ];

  const app = renderer.create(
    <MockedProvider mocks={mockedResponses} addTypename={false}>
      <Recipe navigation={mockedNavigation} route={mockedRoute} />
    </MockedProvider>);

  expect(app).toMatchSnapshot();
});


test('matches snapshot when carbon footprint fails', async () => {
  const mockedResponses = [
    {
      request: {
        query: POST_RECIPE_MUTATION,
      },
      result: () => {
        return {
          data: {
            postRecipe: {
              carbonFootprintPerKg: null,
            },
          },
        };
      },
    },
  ];

  const app = renderer.create(
    <MockedProvider mocks={mockedResponses} addTypename={false}>
      <Recipe navigation={mockedNavigation} route={mockedRoute} />
    </MockedProvider>);

  expect(app).toMatchSnapshot();
});
