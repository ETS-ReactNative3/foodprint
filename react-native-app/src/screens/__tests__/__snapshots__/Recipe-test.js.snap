// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`matches snapshot when carbon footprint fails 1`] = `
<View
  style={
    Object {
      "alignItems": "center",
      "flex": 1,
      "justifyContent": "center",
      "margin": 66.5,
    }
  }
>
  <Image
    resizeMode="contain"
    source={
      Object {
        "testUri": "../../../src/images/full-smiley.png",
      }
    }
    style={
      Object {
        "height": 333.5,
        "marginBottom": 66.5,
      }
    }
  />
  <Text
    style={
      Object {
        "fontSize": 30,
      }
    }
  >
    Give us a recipe name or link, and we'll figure out the carbon footprint.
  </Text>
  <View
    style={
      Object {
        "marginVertical": 26.5,
        "paddingHorizontal": 10,
        "width": "100%",
      }
    }
  >
    <View
      style={
        Object {
          "alignItems": "center",
          "borderBottomWidth": 1,
          "borderColor": "#86939e",
          "flexDirection": "row",
          "transform": Array [
            Object {
              "translateX": 0,
            },
          ],
        }
      }
    >
      <TextInput
        allowFontScaling={true}
        editable={true}
        onChangeText={[Function]}
        onSubmitEditing={[Function]}
        rejectResponderTermination={true}
        style={
          Object {
            "alignSelf": "center",
            "color": "black",
            "flex": 1,
            "fontSize": 18,
            "minHeight": 40,
          }
        }
        testID="RNE__Input__text-input"
        underlineColorAndroid="transparent"
        value=""
      />
    </View>
  </View>
  <View
    style={
      Object {
        "borderRadius": 3,
        "marginVertical": 26.5,
      }
    }
  >
    <TouchableOpacity
      activeOpacity={0.3}
      disabled={false}
      onPress={[Function]}
    >
      <View
        style={
          Object {
            "alignItems": "center",
            "backgroundColor": "green",
            "borderColor": "#2089dc",
            "borderRadius": 3,
            "borderWidth": 0,
            "flexDirection": "row",
            "height": 45,
            "justifyContent": "center",
            "padding": 8,
            "width": 225,
          }
        }
      >
        <Text
          style={
            Object {
              "color": "white",
              "fontSize": 37.5,
              "paddingBottom": 1,
              "paddingTop": 2,
              "textAlign": "center",
            }
          }
        >
          GO
        </Text>
      </View>
    </TouchableOpacity>
  </View>
</View>
`;

exports[`matches snapshot when carbon footprint success 1`] = `
<View
  style={
    Object {
      "alignItems": "center",
      "flex": 1,
      "justifyContent": "center",
      "margin": 66.5,
    }
  }
>
  <Image
    resizeMode="contain"
    source={
      Object {
        "testUri": "../../../src/images/full-smiley.png",
      }
    }
    style={
      Object {
        "height": 333.5,
        "marginBottom": 66.5,
      }
    }
  />
  <Text
    style={
      Object {
        "fontSize": 30,
      }
    }
  >
    Give us a recipe name or link, and we'll figure out the carbon footprint.
  </Text>
  <View
    style={
      Object {
        "marginVertical": 26.5,
        "paddingHorizontal": 10,
        "width": "100%",
      }
    }
  >
    <View
      style={
        Object {
          "alignItems": "center",
          "borderBottomWidth": 1,
          "borderColor": "#86939e",
          "flexDirection": "row",
          "transform": Array [
            Object {
              "translateX": 0,
            },
          ],
        }
      }
    >
      <TextInput
        allowFontScaling={true}
        editable={true}
        onChangeText={[Function]}
        onSubmitEditing={[Function]}
        rejectResponderTermination={true}
        style={
          Object {
            "alignSelf": "center",
            "color": "black",
            "flex": 1,
            "fontSize": 18,
            "minHeight": 40,
          }
        }
        testID="RNE__Input__text-input"
        underlineColorAndroid="transparent"
        value=""
      />
    </View>
  </View>
  <View
    style={
      Object {
        "borderRadius": 3,
        "marginVertical": 26.5,
      }
    }
  >
    <TouchableOpacity
      activeOpacity={0.3}
      disabled={false}
      onPress={[Function]}
    >
      <View
        style={
          Object {
            "alignItems": "center",
            "backgroundColor": "green",
            "borderColor": "#2089dc",
            "borderRadius": 3,
            "borderWidth": 0,
            "flexDirection": "row",
            "height": 45,
            "justifyContent": "center",
            "padding": 8,
            "width": 225,
          }
        }
      >
        <Text
          style={
            Object {
              "color": "white",
              "fontSize": 37.5,
              "paddingBottom": 1,
              "paddingTop": 2,
              "textAlign": "center",
            }
          }
        >
          GO
        </Text>
      </View>
    </TouchableOpacity>
  </View>
</View>
`;
